### PR TITLE
AMP-SHADOW-DOC-LOADER extension for PWA

### PR DESCRIPTION
**Pull request by oliverfernandez**
AMP-SHADOW-DOC-LOADER extensions wraps attachShadowDoc calls in a AMP component, so a PWA can benefit from AMP components life cycle when instantiating new views

### CODE REVIEW

File: examples/pwa/pwa.html [hunk 1]
```diff
@@ -92,6 +92,12 @@
           <div class="detail">Fusce pretium tempor justo, vitae consequat dolor maximus eget.</div>
         </a>
       </article>
+        <article class="card">
+            <a href="/pwa/examples/amp-shadow-doc-loader.amp.max.html">
```
Comment (dvoytenko) [inline examples/pwa/pwa.html]: fyi: we got rid of ".max." thing - they are now all max by default.

File: extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  MANUAL: 'manual'
+};
+
+let uuid = 0;
+
+export default class AmpShadowDocLoader extends AMP.BaseElement {
+  constructor(element) {
+    super(element);
+
+    this.xhr_ = xhrFor(element.ownerDocument.defaultView);
```
Comment (dvoytenko) [inline extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js]: `this.win`

File: extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    this.ampDocRequest_ = null;
+    this.ampDoc_ = null;
+    this.retry_ = this.retry_.bind(this);
+
+    this.registerAction('retry', this.retry_);
```
Comment (dvoytenko) [inline extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js]: Actions should be registered in the `buildCallback`

File: extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+  handleDocRequestError_(error) {
+    this.dispatchEvent_('error', error);
+    this.createRetryButton_();
+  }
+
+  attachAMPDocument_(doc) {
+    return new Promise(resolve => {
+      this.deferMutate(() => {
+        this.removePlaceholder_();
+        this.ampDoc_ = AMP.attachShadowDoc(this.element, doc, this.docUrl_,
```
Comment (dvoytenko) [inline extensions/amp-shadow-doc-loader/0.1/amp-shadow-doc-loader.js]: Let's use streaming option right away.

**Comment by dvoytenko** [comment]
@oliverfernandez Thanks again. I left a few simpler notes. The big thing we need to discuss here is where this extension belongs:

1. Should it be amphtml repo or a separate repo
2. Do we allow it for leaf AMP documents? Or only shell? Possibly, we could make it an non-validateable extension.
3. I feel like, based on this example, we'll have some nuances with viewport. Normally our viewport currently assumes that the doc is roughly takes up the full screen, or at least anchored close to the top of the screen.

Alternatively, if there was a special shell implementation - we could just make it part of the shell binary?

**Comment by oliverfernandez** [comment]
@dvoytenko 

1. This extension only makes sense  when AMP runtime is in shadow mode and we have an `AmpDocShell`, because it's really performant to "hide" the call to `attachShadowDoc` inside an AMP extension. If we want to make `AmpDocShell` optional, then perhaps is better idea to move it to another repository. If not, we could leave it here, and make this extension only available when AMP runtime initializes in shadow mode.

2. Allowing this extension in leaf AMP documents would allow us to have infinite scroll very easy (at the end of a leaf loads the next one without the shell having to do anything). The drawback would be that we could end up with deep nested AmpDoc. Do you think AMP could handle this?

3. You can see a good live example of what you mention [here](https://www.davidemaggio.it/archives/146938/tutto-puo-succedere-2-anticipazioni-dodicesima-puntata) At the end of the article body appears the homepage, which is another `AmpDoc` loaded via an `Amp-Shadow-Doc-Loader` We have seen that everything works fine (extensions, services like analytics and so on), but perhaps you could take a look and check if there's something wrong

My feeling tells me that having always available the `AmpDocShell` is good to expose AMP services, and allows to ensure best practices for PWA's built around AMP

**Comment by dvoytenko** [comment]
@oliverfernandez My sense is that let's focus on #9638 first.

> This extension only makes sense when AMP runtime is in shadow mode and we have an AmpDocShell, because it's really performant to "hide" the call to attachShadowDoc inside an AMP extension. If we want to make AmpDocShell optional, then perhaps is better idea to move it to another repository. If not, we could leave it here, and make this extension only available when AMP runtime initializes in shadow mode.

Let's keep the extension here. But let's guard it by the same experiment that I suggested in #9638.

> Allowing this extension in leaf AMP documents would allow us to have infinite scroll very easy (at the end of a leaf loads the next one without the shell having to do anything). The drawback would be that we could end up with deep nested AmpDoc. Do you think AMP could handle this?

I might be misunderstanding, but this seems to contradict a bit "This extension only makes sense when AMP runtime is in shadow mode". Could you pleas elaborate? Generally, I think, if our goal is to only show the home page - this could work fine. If this would be a infinite scroll - it's more complicated b/c infinite scroll has to also have garbage collection. I'd like to see exactly where the concept of infinite scroll belongs: shell, leaf doc, etc. If we can start with shell - let's do.

**Comment by ericlindley-g** [comment]
I got a question about the status of this PR — is this planned for launch? Any key blockers?

**Comment by aghassemi** [comment]
Thanks for the pull request. Since this PR has been stale for a while, I am closing it. Please feel free to reopen as needed.

**Pull request closed without merging**